### ISSUE TITLE
UI / C++ / Rack question

### ISSUE THREAD

**Issue by alto777** [VCVRack/Rack#845]
If this is not the right place to ask, please smack me with a damp trout and point me away.

A. Can I depend on placing UI objects on top of existing objects just by placing them later in the widget, that is will they come out on top (so to speak) and receive and eat mouse interactions?

B. Is the hit region for a UI object a rectangle in all cases?

Background:

I have a plugin with some snap knobs, but I would like to have them be not-snappy sometimes. While thinking this through, I thought of two solutions.

1. A button/LED per knob showing snap/not snap. OR

2. A little non-snappy knob on top of the knob, like the fine tuning thing you see on old oscilloscopes. Never mind precisely how old.

In (1), I don't have the panel real estate, so I tried placing it on top of the knob, and it works.

In (2), a tiny non-snap knob on top of the bigger snap knob also works.

So now I'm getting greedy, and I tried a stack of big snap knob, little vernier knob and finally a button/LED. Aside from fussing with the coordinates to get them all lined up, it works like a charm.

It looks to me like I can get away with this, with a bit of danger around the rectangle hit region you might not get what you think exactly. I dunno if it's such a great UI thing, we'll see.

**Comment by AndrewBelt** [comment]
A. Yes, widgets are drawn oldest first, so the first widgets are on the bottom.

B. Yes. You can filter this if you really need by programmatically consuming onMouseDown events, but note that users expect bounding boxes to be rectangle.

My question is why you need snapping knobs if the value you're controlling is continuous.

**Comment by alto777** [comment]
Thank you for the rapid and kind response. I understand by your answer that I will be OK, that is the underlying objects will not be activated if the upper object handles a mouse. This is what I would hope and expect.

As to why:

I am working on combining two old synthesizer things I've seen. One depends on selecting pitches chromatically, hence the snap knob for semitones. The other uses weighted sums not locked into any Western prejudices like even temper. So I got the brain wave of have a snap knob with a semitone vernier and started to play around with it.

It is testament more to the organization and clarity of the source code than my own cleverness (such as it may well be) that it was quite easy to hack up a quick test of this idea. By placing the reset value of the vernier at half-rotation I got just what I wanted.

THX again!
